### PR TITLE
Add functionality to mark `used` and `produced` columns as optional.

### PR DESCRIPTION
**Pull request by dsavoiu**
This PR introduces a mechanism for more fine-grained control over which input/output columns of array functions should be covered by the checks introduced in PR #240. In the current implementation, these checks can only be turned on or off for all columns simultaneously, which limits the usability of this feature for functions that only read/produce certain columns under specific circumstances.

By default, columns given  in the `uses` (`produces`) keywords of `ArrayFunction`s are assumed to be required and their presence in the input (output) arrays will be checked . To mark optional columns, an `optional_column` function is provided in `columnar_util`, which can be used as follows:
```python
from columnflow.production import producer
from columnflow.columnar_util import optional_column

@producer(
    uses={
        "required_column",
        optional_column("my_optional_column"),
    },
    produces={
        "always_produced_column",
        optional_column("maybe_produced_column"),
    },
)
def my_producer(...):
    ...
```

### CODE REVIEW

**Review by dsavoiu — COMMENTED** [review]
Neat, thanks for implementing this! Using `Route` with the new tagging mechanism really looks much cleaner and more flexible than the additional `OptFlagged` wrapper.

One general thing I noticed is that the `Route` constructor now always seems to create a copy if passed an existing route object, which I guess is unavoidable since the object state now includes a `tags` field. I'm not sure if we relied on the no-copy behavior anywhere else in the code base, so maybe we should check that, but otherwise I think the additional tagging functionality is well worth the copy overhead.

Apart from that, I only found a few minor things to address below, but otherwise this looks good to merge from my side.

File: columnflow/columnar_util.py [hunk 1]
```diff
@@ -1743,9 +1743,9 @@ def __call__(self, *args, **kwargs):
                 f"{get_source_code(self.skip_func, indent=4)}",
             )
 
-        # assume first argument is event array and
-        # check that the 'used' columns are present
-        if "uses" in self.check_columns_present:
+        # assume result is an event array or tuple with an event array as the first element and
+        # check that the 'produced' columns are present
+        if self.check_produced_columns:
```
[the remark this answers is not in the document]
Comment (riga) [reply columnflow/columnar_util.py]: Indeed! It's fixed now.

File: columnflow/columnar_util.py [hunk 2]
```diff
@@ -1701,28 +1690,39 @@ def _check_columns(self, ak_array: ak.Array, io_flag: IOFlag) -> None:
...
+        def exists(route):
+            route_str = str(route)
+            # trivial check for non-patterns
+            if is_pattern(route_str):
+                for _route in get_ak_routes(ak_array):
+                    if fnmatch.fnmatch(str(_route), route_str):
+                        return True
+                return False
+
+            return has_ak_column(ak_array, route)
```
[the remark this answers is not in the document]
Comment (riga) [reply columnflow/columnar_util.py]: You're right, `select_columns` indeed supports patterns and uses the same syntax as our Route objects. I changed it back to what you did before.

File: columnflow/columnar_util.py [hunk 3]
```diff
@@ -361,12 +354,12 @@ def __hash__(self) -> int:
     def __len__(self) -> int:
         return len(self._fields)
 
-    def __eq__(self, other: Route | Sequence[str] | str) -> bool:
+    def __eq__(self, other: Route | str | Sequence[str | int | slice | type(Ellipsis) | list]) -> bool:
         if isinstance(other, Route):
             return self.fields == other.fields
-        elif isinstance(other, (list, tuple)):
+        if isinstance(other, tuple):
```
[the remark this answers is not in the document]
Comment (riga) [reply columnflow/columnar_util.py]: This is just a residual change that I tested along the way and it should have actually been reverted. It's fixed now.

File: columnflow/columnar_util.py [hunk 4]
```diff
@@ -182,7 +171,7 @@ def slice_to_str(cls, s: slice) -> str:
     def _join(
         cls,
         sep: str,
-        fields: Sequence[str | int | slice | type(Ellipsis) | tuple | list],
+        fields: Sequence[str | int | slice | type(Ellipsis) | list],
```
[the remark this answers is not in the document]
Comment (riga) [reply columnflow/columnar_util.py]: Same as above. I added `tuple` back to the type hints.

File: columnflow/columnar_util.py [hunk 5]
```diff
@@ -1282,6 +1295,20 @@ def my_other_func_init(self: ArrayFunction):
        Flag that can be used in nested dependencies between array functions to denote columns names
        in the :py:attr:`produces` set.
 
+    .. py:classattribute:: check_used_columns
+       type: bool
+
+       A flag that decides whether, during the actual call, the input array should be checked for
+       the existence of all columns defined in :py:attr:`uses`. If a column is missing, an exception
+       is raised.
```
[the remark this answers is not in the document]
Comment (riga) [reply columnflow/columnar_util.py]: Yes, I think it's good write this here explicitly. I added that :+1:

**Review by riga — APPROVED** [review]

**Comment by riga** [comment]
All review comments have been implemented as suggested plus additional tests were added covering the new Route tags. Going to merge.

**Pull request merged**